homeassistant/components/androidtv/__init__.py

```diff
@@ -1 +1,221 @@
+        return None
+    return json_rules
+
+
+async def async_connect_androidtv(
+    hass, config, *, state_detection_rules=None, timeout=30.0
+):
+    """Connect to Android device."""
+    address = f"{config[CONF_HOST]}:{config[CONF_PORT]}"
+
+    adbkey, signer, adb_log = await hass.async_add_executor_job(
+        _setup_androidtv, hass, config
+    )
+
+    aftv = await setup(
+        config[CONF_HOST],
+        config[CONF_PORT],
+        adbkey,
+        config.get(CONF_ADB_SERVER_IP),
+        config.get(CONF_ADB_SERVER_PORT, DEFAULT_ADB_SERVER_PORT),
+        state_detection_rules,
+        config[CONF_DEVICE_CLASS],
+        timeout,
+        signer,
+    )
+
+    if not aftv.available:
+        # Determine the name that will be used for the device in the log
+        if config[CONF_DEVICE_CLASS] == DEVICE_ANDROIDTV:
+            device_name = "Android TV device"
+        elif config[CONF_DEVICE_CLASS] == DEVICE_FIRETV:
+            device_name = "Fire TV device"
+        else:
+            device_name = "Android TV / Fire TV device"
+
+        _LOGGER.warning(
+            "Could not connect to %s at %s %s", device_name, address, adb_log
+        )
+        return None
+
+    return aftv
+
+
+def _migrate_aftv_entity(hass, aftv, entry_unique_id):
+    """Migrate a entity to new unique id."""
+    entity_reg = er.async_get(hass)
+
+    entity_unique_id = entry_unique_id
+    if entity_reg.async_get_entity_id(MP_DOMAIN, DOMAIN, entity_unique_id):
+        # entity already exist, nothing to do
+        return
+
+    old_unique_id = aftv.device_properties.get(PROP_SERIALNO)
+    if not old_unique_id:
+        # serial no not found, exit
+        return
+
+    migr_entity = entity_reg.async_get_entity_id(MP_DOMAIN, DOMAIN, old_unique_id)
+    if not migr_entity:
+        # old entity not found, exit
+        return
+
+    try:
+        entity_reg.async_update_entity(migr_entity, new_unique_id=entity_unique_id)
+    except ValueError as exp:
+        _LOGGER.warning("Migration of old entity failed: %s", exp)
+
+
+async def async_setup(hass, config):
+    """Set up the Android TV integration."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Android TV platform."""
+
+    state_det_rules = entry.options.get(CONF_STATE_DETECTION_RULES)
+    json_rules = validate_state_det_rules(state_det_rules)
+
+    aftv = await async_connect_androidtv(
+        hass, entry.data, state_detection_rules=json_rules
+    )
+    if not aftv:
+        raise ConfigEntryNotReady()
+
+    # migrate existing entity to new unique ID
+    if entry.source == SOURCE_IMPORT:
+        _migrate_aftv_entity(hass, aftv, entry.unique_id)
+
+    async def async_close_connection(event):
+        """Close Android TV connection on HA Stop."""
+        await aftv.adb_close()
+
+    stop_listener = hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STOP, async_close_connection
+    )
```

Comment:
You can wrap this in `entry.async_on_unload` instead.  Same with the `add_update_listener`.  Then you don't have to store it in `hass.data`